jest.config.ts

```diff
@@ -36,60 +27,115 @@ const BALANCE_RESULTS_PATH = path.resolve(
+  const tests = new Map<string, number>();
+  const SUITE_P50_DURATION_MS = 1500;
+
+  // First, iterate over all of the tests we have stats for.
+  for (const test in testStats) {
+    if (testStats[test] <= 0) {
+      throw new Error(`Test duration is <= 0 for ${test}`);
+    }
+    tests.set(test, testStats[test]);
+  }
+  // Then, iterate over all of the remaining tests and assign them a default duration.
+  for (const test of allTests) {
+    if (tests.has(test)) {
+      continue;
     }
+    tests.set(test, SUITE_P50_DURATION_MS);
+  }
 
-    return index;
+  // Sanity check to ensure that we have all of our tests accounted for, we need to fail
+  // if this is not the case as we risk not executing some tests and passing the build.
+  if (tests.size < allTests.length) {
+    throw new Error(
+      `All tests should be accounted for, missing ${allTests.length - tests.size}`
+    );
   }
 
-  /**
-   * We may not have a duration for all tests (e.g. a test that was just added)
-   * as the `source` needs to be generated
-   */
-  for (const test of allTests) {
-    const index = findSmallestGroup();
-    results.set(test, index);
+  const groups: string[][] = [];
+
+  // We sort files by path so that we try and improve the transformer cache hit rate.
+  // Colocated domain specific files are likely to require other domain specific files.
+  const testsSortedByPath = Array.from(tests.entries()).sort((a, b) =>
+    a[0].localeCompare(b[0])
+  );
+
+  for (let group = 0; group < nodeTotal; group++) {
+    groups[group] = [];
+    let duration = 0;
 
-    if (source[test] !== undefined) {
-      totalRunTimes[index] = totalRunTimes[index] + source[test];
+    // While we are under our target duration and there are tests in the group
+    while (duration < targetDuration && testsSortedByPath.length > 0) {
+      // We peek the next item to check that it is not some super long running
+      // test that may exceed our target duration. For example, if target runtime for each group is
+      // 10 seconds, we have currently accounted for 9 seconds, and the next test is 5 seconds, we
+      // want to move that test to the next group so as to avoid a 40% imbalance.
+      const peek = testsSortedByPath[testsSortedByPath.length - 1];
+      if (duration + peek[1] > targetDuration && peek[1] > 30_000) {
+        break;
+      }
+      const nextTest = testsSortedByPath.pop();
+      if (!nextTest) {
+        throw new TypeError('Received falsy test' + JSON.stringify(nextTest));
+      }
+      groups[group].push(nextTest[0]);
+      duration += nextTest[1];
     }
   }
 
-  return results;
+  // Whatever may be left over will get round robin'd into the groups.
+  const i = 0;
+  while (testsSortedByPath.length) {
+    const nextTest = testsSortedByPath.pop();
+    if (!nextTest) {
+      throw new TypeError('Received falsy test' + JSON.stringify(nextTest));
+    }
+    groups[i % 4].push(nextTest[0]);
+  }
```

Comment:
`i` is always 0, you're always pushing to the same group, no?

Reply:
Arf, yes. Nice find